### ISSUE TITLE
[Pytorch-Upstream] Softmax backward is slow

### ISSUE THREAD

**Issue by orrangetabby17** [intel/intel-xpu-backend-for-triton#7610]
### Describe the issue

we found that softmax backward related fusion kernel is slow on PVC(274.08GB/s).

On PVC the triton kernel is:
```python
import triton
import triton.language as tl

from torch._inductor.runtime import triton_helpers, triton_heuristics
from torch._inductor.runtime.triton_helpers import libdevice, math as tl_math
from torch._inductor.runtime.hints import AutotuneHint, ReductionHint, TileHint, DeviceProperties
triton_helpers.set_driver_to_gpu()

from torch._dynamo.testing import rand_strided
from torch._C import _xpu_getCurrentRawStream as get_raw_stream
import torch

@triton_heuristics.reduction(
    size_hints={'x': 65536, 'r0_': 1024},
    reduction_hint=ReductionHint.INNER,
    filename=__file__,
    triton_meta={'signature': {'in_out_ptr0': '*fp32', 'in_ptr0': '*fp32', 'in_ptr1': '*fp32', 'in_ptr2': '*fp32', 'in_ptr3': '*fp32', 'in_ptr4': '*fp32', 'xnumel': 'i32', 'r0_numel': 'i32', 'XBLOCK': 'constexpr', 'R0_BLOCK': 'constexpr'}, 'device': DeviceProperties(type='xpu', index=0, multi_processor_count=64, cc={'architecture': 13136561920, 'device_id': 3029, 'driver_version': '1.6.33578+51', 'gpu_eu_count': 512, 'gpu_subslice_count': 64, 'has_atomic64': True, 'has_bfloat16_conversions': True, 'has_fp16': True, 'has_fp64': True, 'has_subgroup_2d_block_io': True, 'has_subgroup_matrix_multiply_accumulate': True, 'has_subgroup_matrix_multiply_accumulate_tensor_float32': False, 'is_integrated_gpu': False, 'last_level_cache_size': 201326592, 'local_mem_size': 131072, 'max_compute_units': 512, 'max_num_sub_groups': 64, 'max_work_group_size': 1024, 'memory_bus_width': 64, 'memory_clock_rate': 3200, 'name': 'Intel(R) Data Center GPU Max 1550', 'platform_name': 'Intel(R) oneAPI Unified Runtime over Level-Zero', 'sub_group_sizes': [16, 32], 'total_memory': 68702699520, 'type': 'gpu', 'vendor': 'Intel(R) Corporation', 'version': '12.60.7'}, major=None, regs_per_multiprocessor=None, max_threads_per_multi_processor=None, max_threads_per_block=1024, warp_size=None), 'constants': {}, 'native_matmul': False, 'enable_fp_fusion': True, 'launch_pdl': False, 'disable_ftz': False, 'configs': [{(0,): [['tt.divisibility', 16]], (1,): [['tt.divisibility', 16]], (2,): [['tt.divisibility', 16]], (3,): [['tt.divisibility', 16]], (4,): [['tt.divisibility', 16]], (5,): [['tt.divisibility', 16]], (6,): [['tt.divisibility', 16]], (7,): [['tt.divisibility', 16]]}]},
    inductor_meta={'grid_type': 'Grid1D', 'kernel_name': 'triton_red_fused__softmax__softmax_backward_data__unsafe_view_add_permute_view_48', 'mutated_arg_names': ['in_out_ptr0'], 'optimize_mem': True, 'no_x_dim': False, 'atomic_add_found': False, 'num_load': 7, 'num_store': 2, 'num_reduction': 1, 'autotune_hints': set(), 'tiling_scores': {'x': 294976, 'r0_': 169869312}, 'kernel_num_gb': 0.3400336, 'kernel_flop': 0, 'backend_hash': '3F02FCDB6319878ECD32540DD4F73245AD52AD4AD0472E826D0D8343F980724E', 'assert_indirect_indexing': True, 'autotune_local_cache': True, 'autotune_pointwise': True, 'autotune_remote_cache': None, 'force_disable_caches': True, 'dynamic_scale_rblock': True, 'incremental_autotune': False, 'max_autotune': False, 'max_autotune_pointwise': False, 'min_split_scan_rblock': 256, 'spill_threshold': 16, 'store_cubin': False, 'deterministic': False, 'batch_invariant': False, 'force_filter_reduction_configs': False, 'mix_order_reduction_allow_multi_stages': True, 'dynamic_disable_pipelining': True, 'are_deterministic_algorithms_enabled': False}
)
@triton.jit
def triton_red_fused__softmax__softmax_backward_data__unsafe_view_add_permute_view_48(in_out_ptr0, in_ptr0, in_ptr1, in_ptr2, in_ptr3, in_ptr4, xnumel, r0_numel, XBLOCK : tl.constexpr, R0_BLOCK : tl.constexpr):
    xnumel = 36864
    r0_numel = 576
    rnumel = r0_numel
    RBLOCK: tl.constexpr = R0_BLOCK
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:, None]
    xmask = tl.full([XBLOCK], True, tl.int1)[:, None]
    r0_base = tl.arange(0, R0_BLOCK)[None, :]
    rbase = r0_base
    x0 = xindex
    tmp2 = tl.load(in_ptr2 + (((x0 // 576) % 16)), None, eviction_policy='evict_last')
    tmp4 = tl.load(in_ptr3 + (x0), None, eviction_policy='evict_last')
    tmp7 = tl.load(in_ptr4 + (x0), None, eviction_policy='evict_last')
    _tmp11 = tl.full([XBLOCK, R0_BLOCK], 0, tl.float32)
    for r0_offset in tl.range(0, r0_numel, R0_BLOCK):
        r0_index = r0_offset + r0_base
        r0_mask = r0_index < r0_numel
        roffset = r0_offset
        rindex = r0_index
        r0_1 = r0_index
        tmp0 = tl.load(in_ptr0 + (16*r0_1 + 9216*((x0 % 576)) + 5308416*(x0 // 9216) + (((x0 // 576) % 16))), r0_mask, eviction_policy='evict_last', other=0.0)
        tmp1 = tl.load(in_ptr1 + (16*r0_1 + 9216*((x0 % 576)) + 5308416*(x0 // 9216) + (((x0 // 576) % 16))), r0_mask, eviction_policy='evict_last', other=0.0)
        tmp3 = tmp1 + tmp2
        tmp5 = tmp3 - tmp4
        tmp6 = libdevice.exp(tmp5)
        tmp8 = (tmp6 / tmp7)
        tmp9 = tmp0 * tmp8
        tmp10 = tl.broadcast_to(tmp9, [XBLOCK, R0_BLOCK])
        tmp12 = _tmp11 + tmp10
        _tmp11 = tl.where(r0_mask, tmp12, _tmp11)
        tl.store(in_out_ptr0 + (r0_1 + 576*x0), tmp9, r0_mask)
    tmp11 = tl.sum(_tmp11, 1)[:, None]
    for r0_offset in tl.range(0, r0_numel, R0_BLOCK):
        r0_index = r0_offset + r0_base
        r0_mask = r0_index < r0_numel
        roffset = r0_offset
        rindex = r0_index
        r0_1 = r0_index
        tmp13 = tl.load(in_ptr1 + (16*r0_1 + 9216*((x0 % 576)) + 5308416*(x0 // 9216) + (((x0 // 576) % 16))), r0_mask, eviction_policy='evict_last', other=0.0)
        tmp19 = tl.load(in_out_ptr0 + (r0_1 + 576*x0), r0_mask, eviction_policy='evict_first', other=0.0)
        tmp14 = tmp13 + tmp2
        tmp15 = tmp14 - tmp4
        tmp16 = libdevice.exp(tmp15)
        tmp17 = (tmp16 / tmp7)
        tmp18 = -tmp17
        tmp20 = tl.fma(tmp18, tmp11, tmp19)
        tl.store(in_out_ptr0 + (r0_1 + 576*x0), tmp20, r0_mask)


def get_args():
    arg_0 = rand_strided((4, 16, 576, 576), (5308416, 331776, 576, 1), device='xpu:0', dtype=torch.float32)
    arg_1 = rand_strided((1327104, 16), (16, 1), device='xpu:0', dtype=torch.float32)
    arg_2 = rand_strided((1327104, 16), (16, 1), device='xpu:0', dtype=torch.float32)
    arg_3 = rand_strided((16,), (1,), device='xpu:0', dtype=torch.float32)
    arg_4 = rand_strided((4, 16, 576, 1), (9216, 576, 1, 1), device='xpu:0', dtype=torch.float32)
    arg_5 = rand_strided((4, 16, 576, 1), (9216, 576, 1, 1), device='xpu:0', dtype=torch.float32)
    return arg_0, arg_1, arg_2, arg_3, arg_4, arg_5, 36864, 576,


def call(args):
    with torch.xpu._DeviceGuard(0):
        torch.xpu.set_device(0)
        raw_stream0 = get_raw_stream(0)
        triton_red_fused__softmax__softmax_backward_data__unsafe_view_add_permute_view_48.run(*args, stream=raw_stream0)


def benchmark_all_configs(args):
    with torch.xpu._DeviceGuard(0):
        torch.xpu.set_device(0)
        return triton_red_fused__softmax__softmax_backward_data__unsafe_view_add_permute_view_48.benchmark_all_configs(*args)


if __name__ == '__main__':
    from torch._inductor.runtime.benchmarking import benchmarker

    args = get_args()
    ms = benchmarker.benchmark(lambda: call(args), device='xpu', rep=40)
    num_gb = 0.3400336
    gb_per_s = num_gb / (ms / 1e3)
    print(f"{ms:.3f}ms    {num_gb:.3f}GB    {gb_per_s:.2f}GB/s")

```

### Environment details

Triton: triton-xpu         3.7.2+git5fcc14d9
GPU: PVC

**Comment by Niz13** [comment]
## Root cause: non-coalesced (stride-16) loads — fix belongs in inductor tiling

Reproduced on PVC (Max 1550, tile 0): **~247 GB/s**, matching the report.

### Why it's slow

The data is head-innermost (permuted) layout, and inductor folds the only contiguous axis — `head` (16 wide, stride 1) — into the flattened program id via `permute`+`unsafe_view`. Decoding `x0 = b*9216 + h*576 + s`, the load offset is:

```
in_ptr0/in_ptr1 offset = 16*k + 9216*s + 5308416*b + h
                          (k:16) (s:9216) (b:5308416) (h:1, but fixed per program)
```

- `h` is stride-1 but **loop-invariant per program**; both tensor axes (`k` stride 16, `x` stride ~9216) are non-contiguous.
- So every `in_ptr0`/`in_ptr1` read is a **≥64 B strided scalar load** — no vectorization, no 2D block I/O. `in_ptr1` is read twice.
- Only the contiguous store coalesces. ~247 GB/s ≈ 4 B used per 64 B line.

### Evidence (A/B — identical math, buffers, output)

Restructuring so the 16 heads map to the 16 subgroup lanes → coalesced 64 B reads:

| Variant | Time | Bandwidth |
| --- | --- | --- |
| Original | 1.38 ms | **247 GB/s** |
| Head-on-lanes | 0.46 ms | **740 GB/s (3.0×)** |

Correctness `rel = 2e-7`, `allclose = True` (the full-scale `allclose=False` is just fp32 reduction-order noise on values up to 8.5e6).

### Why inductor, not the XPU backend

- The Coalesce pass binds lanes to a **contiguous tensor axis** — and it works here (the store coalesces).
- The slow loads have **no contiguous tensor axis**: the only one (`head`) was flattened into `(x0//576)%16`. `tiling_scores` (`x: 294976`, `r0_: 169869312`) confirm inductor treated all 36864 iterations as flat `x`.
- Recovering `head` means re-tiling and **changing the launch grid** (36864 → 2304 programs) — a codegen job, not a layout pass.
- In the experiment I only exposed `head` as a `tl.arange(0, 16)` axis; the Coalesce pass mapped it to lanes on its own → the backend does the right thing the moment the axis exists.

### Recommendation

Inductor tiling should split the stride-1 `head` axis (size 16) out of the flattened `x` and tile it as the vectorized dimension (maps exactly to XPU's 16-wide subgroup). Repro + A/B scripts available.

[softmax_bwd_repro.py](https://github.com/user-attachments/files/31114085/softmax_bwd_repro.py)
[softmax_bwd_headvec.py](https://github.com/user-attachments/files/31114081/softmax_bwd_headvec.py)

@vlad-penkin how do we move on from here?

**Comment by Niz13** [comment]
https://jira.devtools.intel.com/browse/PYTORCHDGQ-9790

**Comment by Niz13** [comment]
performance is fixed by the PyTorch community https://github.com/pytorch/pytorch/pull/193158
the PyTorch team in the ticket confirmed the performance is fixed with that
